chain/network/src/peer_manager/peer_manager_actor.rs

```diff
@@ -936,64 +890,40 @@ impl PeerManagerActor {
         self.whitelist_nodes.iter().any(|wn| wn.addr.ip() == *ip)
     }
 
-    /// Returns single random peer with close to the highest height
+    /// Returns peers close to the highest height
     fn highest_height_peers(&self) -> Vec<FullPeerInfo> {
-        let connected_peers = self.state.connected_peers.read();
+        let infos: Vec<_> =
+            self.state.connected_peers.read().values().map(|p| p.full_peer_info()).collect();
+
         // This finds max height among peers, and returns one peer close to such height.
-        let max_height = match (connected_peers.values())
-            .map(|connected_peer| connected_peer.full_peer_info.chain_info.height)
-            .max()
-        {
+        let max_height = match infos.iter().map(|i| i.chain_info.height).max() {
             Some(height) => height,
             None => return vec![],
         };
         // Find all peers whose height is within `highest_peer_horizon` from max height peer(s).
-        connected_peers
-            .values()
-            .filter(|cp| {
-                cp.full_peer_info.chain_info.height.saturating_add(self.config.highest_peer_horizon)
-                    >= max_height
+        infos
+            .into_iter()
+            .filter(|i| {
+                i.chain_info.height.saturating_add(self.config.highest_peer_horizon) >= max_height
             })
-            .map(|cp| cp.full_peer_info.clone())
-            .collect::<Vec<_>>()
+            .collect()
     }
 
     // Get peers that are potentially unreliable and we should avoid routing messages through them.
     // Currently we're picking the peers that are too much behind (in comparison to us).
     fn unreliable_peers(&self) -> HashSet<PeerId> {
         let my_height = self.state.chain_info.load().height;
-
-        let connected_peers = self.state.connected_peers.read();
         // Find all peers whose height is below `highest_peer_horizon` from max height peer(s).
-        connected_peers
+        self.state
+            .connected_peers
+            .read()
             .values()
-            .filter(|cp| {
-                cp.full_peer_info.chain_info.height.saturating_add(UNRELIABLE_PEER_HORIZON)
+            .filter(|p| {
+                p.chain_height.load(Ordering::Relaxed).saturating_add(UNRELIABLE_PEER_HORIZON)
```

Comment:
let's do this `.load` once

Reply:
wdym? It is loaded once per peer and once for the current node.

Reply:
Nevermind, I can't read, thought that that was `my_height`.